pkg/event/registry.go

```diff
@@ -0,0 +1,163 @@
+package event
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/submariner-io/admiral/pkg/log"
+	k8sV1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+)
+
+type Registry struct {
+	eventHandlers []Handler
+	name          string
+	networkPlugin string
+}
+
+func NewRegistry(name, networkPlugin string) Registry {
+	return Registry{
+		name: name,
+		networkPlugin: networkPlugin,
+	}
+}
+
+// Adds a supported driver, prints a fatal error in the case of double registration
+func (er *Registry) AddHandler(eventHandler Handler) {
+	evNetworkPlugin := eventHandler.GetNetworkPlugin()
+
+	if evNetworkPlugin == AnyNetworkPlugin || evNetworkPlugin == er.networkPlugin {
+		er.eventHandlers = append(er.eventHandlers, eventHandler)
+		klog.Infof("Event handler %q loaded for the %q registry ",
+			eventHandler.GetName(), er.name)
+	} else {
+		klog.V(log.DEBUG).Infof("Event handler %q ignored for the %q registry ",
+			eventHandler.GetName(), er.name)
+	}
```

Comment:
> I'm not clear of the purpose of the `networkPlugin` field defined on the `Registry`. It seems then we'd need a registry of network plugin `Registry`'s? How will handlers know to get the right `Registry` to register with? I'd think there would be one `Registry` for all handlers.

I guess that my intent will be clearer when I finish the unit tests.

The idea was... that .. on the main of the routeagent/networksyncer they will create a registry, and state which network plugin is being used:

```
handlerRegistry := NewRegistry("route-agent", $NETWORK_PLUGIN_FROM_ENV_this_is_given_by_operator)
handlerRegistry.AddHandler(logger.NewHandler())
handlerRegistry.AddHandler(kubeproxy_iptables_handler.NewHandler())
handlerRegistry.AddHandler(ovn_routeagent_handler.NewHandler())
```

Then later on we can create the controller:
```
controller := controller.NewController(hanlderRegistry)
controller.Start(stopCh)
```


But the idea is that the registry already knows on-runtime, which network plugin are we serving for, so it knows which handlers it can discard (dynamically)... instead of keeping them all and returning a list afterwards, I thought we could just optimize that memory and free it as soon as possible. (not that it's a lot.., but won't be used)

Reply:
OK - that clarifies it. There will be a singleton registry which will filter out handlers based on process input.

Reply:
Correct, I will be updating this patch with full testing in a second.